# Support Vector Machines (SVM)
---
As we read earlier, the Support Vector Machine is a type of classification-based, supervised Machine Learning algorithm. In this tutorial, we will learn how to implement a SVM model using Scikit Learn.  

## Importing Project Dependencies
---
Let us first import all the necesary libraries.

In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

## Importing  & Pre-Processing the Dataset
---
For this project, we will be working on sklearn's built-in breast cancer dataset. This is a binary classification problem where on the basis of the given features, we want our model to predict whether a given tumore is malignant (cancerous, denoted by 1) or benign (non-cancerous, denoted by 0).

Let us import the dataset now.

In [2]:
# Step 1- Importing the dataset class
from sklearn.datasets import load_breast_cancer

# Step 2- Creating the dataset class object
cancer = load_breast_cancer()

Now, let us have a look at what all info the breast cancer dataset comes with.

In [3]:
cancer.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [4]:
# printing the dataset description
print(cancer['DESCR'])

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

If you go through the description of the dataset, you will find that there are no null values within the dataset, which means it is already cleaned. Now, let us have a look at the features within our dataset.

In [5]:
# features within our dataset
cancer['feature_names']

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

Now, we will convert our dataset into a pandas dataframe object that will allow us to manipulate and work with the data in a much better way as compared to a numpy array. We will also have a look at the basic info regarding the data.

In [6]:
feature_df = pd.DataFrame(cancer['data'], columns = cancer['feature_names']) # creating the feature dataframe
feature_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [7]:
target = cancer['target'] # target values
target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

Now, before we do the modelling, let us have a look at the our data.

In [8]:
feature_df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

If we compare the data in different columns of our feature set, we will notice that the values vary a lot in terms of scale. In order to tackle this data imbalance, we will have to scale/normalize the data. We will be using MinMaxScaler for this, and what it does is that it reduces all the values in the range \[0,1]. Let us see how to implement that.

In [9]:
#Step 1- Importing MinMaxScaler class
from sklearn.preprocessing import MinMaxScaler

# Step 2- Creating class object
scaler = MinMaxScaler()

# Step 3- Fitting and transforming the feature set
X_scaled = scaler.fit_transform(feature_df)

Now, let us split our dataset into training set and test set.

In [10]:
# Step 1- Importing the test_train_split method from sklearn
from sklearn.model_selection import train_test_split

# Step 2- Performing the split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, target, test_size=0.2, random_state=42)

X_train.shape[0], X_test.shape[0]

(455, 114)

## Modeling
---
Now that the data is ready and preprocessed, we will now train our SVM model.

In [11]:
# Step 1- Importing the support vector classifier class
from sklearn.svm import SVC

# Step 2- Creating the class object
model = SVC()

# Step 3- Training the model
model.fit(X_train,y_train)

SVC()

Now, with the model use our trained model to perform inference on the test set.

In [12]:
# generating predictions on the test set
y_preds = model.predict(X_test)
y_preds

array([1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1,
       0, 1, 1, 0])

Now, let us test the evaluate our model on different accuracy measures.

In [13]:
# importing different accuracy metrics
from sklearn.metrics import classification_report,confusion_matrix

In [14]:
print(confusion_matrix(y_test, y_preds))

[[41  2]
 [ 1 70]]


As we can see, out model performed splendidly, with only 4 false positives and 0 false negative predictions.

In [15]:
pd.DataFrame(classification_report(y_test, y_preds, output_dict = True))

0          1  accuracy   macro avg  weighted avg
precision   0.976190   0.972222  0.973684    0.974206      0.973719
recall      0.953488   0.985915  0.973684    0.969702      0.973684
f1-score    0.964706   0.979021  0.973684    0.971863      0.973621
support    43.000000  71.000000  0.973684  114.000000    114.000000

As we can see, our model has an accuracy of 96.5%, f1 score of 0.97 a precision score of 1 and a recall of 0.9. 

__For cases like breast cancer prediction (and other medical uses of machine learning), we generally prefer the precision over other metrics.__ This is because we want our model to have as less false negatives as possible. In other words, telling a patient with cancer that they don't have cancer can be fatal for them. On the other hand, a patient who doesn't have cancer being told that they have cancer is not that big a problem. They can have tests done later to confirm that they don't have cancer.

As we can see, out model had a precision of 1 (100%), hence it performed really well.

With this, we come at the end of this tutorial. Go through the notebook once again before moving on to the quiz in the next step.